### PR TITLE
HHH-14196 Add parsing of persistence.xml/orm.xml documents in the EE 9 namespace

### PR DESCRIPTION
**Pull request by scottmarlow**
https://hibernate.atlassian.net/browse/HHH-14196

The idea is that Jakarta EE 9 persistence schemas can be parsed/handled in Hibernate ORM versions that deal with Jakarta EE 8.  This will help Jakarta EE 9 implementations such as https://github.com/bstansberry/wildfly/tree/ee-9

As noted in HHH-14196, this pr also added code for Jakarta EE 8 (JPA 2.2) XSDs which might of been missing (not really sure).

Could also use help in understanding if org.hibernate.jpamodelgen.xml.JpaDescriptorParser should reference additional schemas than just JPA 2.1?  I'm not sure if that module has a way to know which JPA version schemas to use, but am curious.  From the Jakarta EE 9 Platform requirements, it is okay to still reference the EE 8 schemas (JPA 2.1 or JPA 2.2).

During testing of https://github.com/bstansberry/wildfly/tree/ee-9 with the Jakarta EE 9 Platform TCK, we see failures like:
```
caused by: org.hibernate.boot.InvalidMappingException: Could not parse mapping document: /content/jpa_core_annotations_elementcollection_vehicles.ear/lib/jpa_core_annotations_elementcollection.jar/myMappingFile.xml (INPUT_STREAM)
	at org.hibernate.boot.jaxb.internal.InputStreamXmlSource.doBind(InputStreamXmlSource.java:46)
	at org.hibernate.boot.jaxb.internal.InputStreamXmlSource.doBind(InputStreamXmlSource.java:38)
	at org.hibernate.boot.spi.XmlMappingBinderAccess.bind(XmlMappingBinderAccess.java:79)
	at org.hibernate.boot.model.process.internal.ScanningCoordinator.applyScanResultsToManagedResources(ScanningCoordinator.java:207)
	at org.hibernate.boot.model.process.internal.ScanningCoordinator.coordinateScan(ScanningCoordinator.java:82)
	at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.prepare(MetadataBuildingProcess.java:99)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:240)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:168)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:32)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:89)
	at org.jboss.as.jpa.hibernate5.TwoPhaseBootstrapImpl.<init>(TwoPhaseBootstrapImpl.java:39)
	at org.jboss.as.jpa.hibernate5.HibernatePersistenceProviderAdaptor.getBootstrap(HibernatePersistenceProviderAdaptor.java:252)
	at org.jboss.as.jpa.service.PhaseOnePersistenceUnitServiceImpl.createContainerEntityManagerFactoryBuilder(PhaseOnePersistenceUnitServiceImpl.java:254) [wildfly-jpa-22.0.0.Beta1-SNAPSHOT.jar:22.0.0.Beta1-SNAPSHOT]
	at org.jboss.as.jpa.service.PhaseOnePersistenceUnitServiceImpl.access$900(PhaseOnePersistenceUnitServiceImpl.java:59) [wildfly-jpa-22.0.0.Beta1-SNAPSHOT.jar:22.0.0.Beta1-SNAPSHOT]
	at org.jboss.as.jpa.service.PhaseOnePersistenceUnitServiceImpl$1$1.run(PhaseOnePersistenceUnitServiceImpl.java:125) [wildfly-jpa-22.0.0.Beta1-SNAPSHOT.jar:22.0.0.Beta1-SNAPSHOT]
	... 10 more
Caused by: org.hibernate.boot.UnsupportedOrmXsdVersionException: Encountered unsupported orm.xml xsd version [3.0] : origin(/content/jpa_core_annotations_elementcollection_vehicles.ear/lib/jpa_core_annotations_elementcollection.jar/myMappingFile.xml)
	at org.hibernate.boot.jaxb.internal.MappingBinder.doBind(MappingBinder.java:74)
	at org.hibernate.boot.jaxb.internal.AbstractBinder.doBind(AbstractBinder.java:102)
	at org.hibernate.boot.jaxb.internal.AbstractBinder.bind(AbstractBinder.java:57)
	at org.hibernate.boot.jaxb.internal.InputStreamXmlSource.doBind(InputStreamXmlSource.java:43)
	... 24 more
```
And

```
jakarta.persistence.PersistenceException: Unable to locate persistence units
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:85)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:61)
	at org.hibernate.jpa.HibernatePersistenceProvider.generateSchema(HibernatePersistenceProvider.java:156)
	at jakarta.persistence.Persistence.generateSchema(Persistence.java:115)
	at com.sun.ts.tests.jpa.jpa22.datetime.Client.setup(Client.java:65)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ts.lib.harness.EETest.run(EETest.java:569)
	at com.sun.ts.lib.harness.ServiceEETest.run(ServiceEETest.java:115)
	at com.sun.ts.tests.common.vehicle.puservlet.PUServletVehicle.runTest(PUServletVehicle.java:65)
	at com.sun.ts.tests.common.vehicle.servlet.ServletVehicle.doGet(ServletVehicle.java:88)
	at com.sun.ts.tests.common.vehicle.servlet.ServletVehicle.doPost(ServletVehicle.java:108)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:520)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.elytron.web.undertow.server.ElytronRunAsHandler.lambda$handleRequest$1(ElytronRunAsHandler.java:68)
	at org.wildfly.security.auth.server.FlexibleIdentityAssociation.runAsFunctionEx(FlexibleIdentityAssociation.java:103)
	at org.wildfly.security.auth.server.Scoped.runAsFunctionEx(Scoped.java:161)
	at org.wildfly.security.auth.server.Scoped.runAs(Scoped.java:73)
	at org.wildfly.elytron.web.undertow.server.ElytronRunAsHandler.handleRequest(ElytronRunAsHandler.java:67)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at org.wildfly.elytron.web.undertow.server.servlet.CleanUpHandler.handleRequest(CleanUpHandler.java:38)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.SendErrorPageHandler.handleRequest(SendErrorPageHandler.java:52)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.SessionRestoringHandler.handleRequest(SessionRestoringHandler.java:119)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1530)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1530)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1530)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1530)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1530)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:387)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:841)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at org.xnio.XnioWorker$WorkerThreadFactory$1$1.run(XnioWorker.java:1280)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Unrecognized JPA persistence.xml XSD version : `3.0`
	at org.hibernate.boot.xsd.ConfigXsdSupport.jpaXsd(ConfigXsdSupport.java:79)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.validate(PersistenceXmlParser.java:467)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.loadUrl(PersistenceXmlParser.java:439)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.parsePersistenceXml(PersistenceXmlParser.java:242)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.parsePersistenceXml(PersistenceXmlParser.java:235)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.doResolve(PersistenceXmlParser.java:229)
	at org.hibernate.jpa.boot.internal.PersistenceXmlParser.locatePersistenceUnits(PersistenceXmlParser.java:67)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:81)
	... 60 more
```

### CODE REVIEW

**Comment by scottmarlow** [comment]
I need to make another orm.xml 3.0 parsing change before this is merged.

**Comment by scottmarlow** [comment]
orm.xml 3.0 parsing is looking better now.

**Comment by Sanne** [comment]
Hi @scottmarlow , had a quick look and it looks good - but a test is failing. Could you investigate please?

File: hibernate-core/src/main/java/org/hibernate/cfg/EJB3DTDEntityResolver.java [hunk 1]
```diff
@@ -41,13 +41,27 @@ public boolean isResolved() {
...
+			}
+			else if ( systemId.endsWith( "orm_2_1.xsd" ) ) {
 				InputStream dtdStream = getStreamFromClasspath( "orm_2_1.xsd" );
 				final InputSource source = buildInputSource( publicId, systemId, dtdStream, false );
 				if ( source != null ) {
 					return source;
 				}
 			}
+			else if ( systemId.endsWith( "orm_2_2.xsd" ) ) {
+				InputStream dtdStream = getStreamFromClasspath( "orm_2_2.xsd" );
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/cfg/EJB3DTDEntityResolver.java:59]: Was something missing here for JPA 2.2 ?

Comment (scottmarlow) [reply hibernate-core/src/main/java/org/hibernate/cfg/EJB3DTDEntityResolver.java:59]: I didn't see an entry for http://www.oracle.com/webfolder/technetwork/jsc/xml/ns/persistence/orm_2_2.xsd which I think was added for https://jakarta.ee/specifications/persistence/2.2.  Note that the Jakarta EE 8 Platform TCK did not update to the 2.2 schemas but some applications may of.

**Comment by scottmarlow** [comment]
> Hi @scottmarlow , had a quick look and it looks good - but a test is failing. Could you investigate please?

Thanks for catching that!  I updated the hibernate-core/src/test/resources/org/hibernate/test/annotations/xml/ejb3/orm5.xml b/hibernate-core/src/test/resources/org/hibernate/test/annotations/xml/ejb3/orm5.xml to use `version="2020"` as the invalid orm schema version.  :-)

File: hibernate-core/src/main/java/org/hibernate/cfg/EJB3DTDEntityResolver.java (hunk 1 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/cfg/EJB3DTDEntityResolver.java:59]: ok. Would you say we need to track it as a separate bugfix?

**Comment by Sanne** [comment]
Thanks @scottmarlow , merging this and including in branches `master` (v. 5.5x), `5.4` and `5.3`.

**Comment by aligit** [comment]
Hi @scottmarlow and thanks for your help.
I switched to 6.0.0.Alpha6 in order to migrate to jakarta 9. Do you have any suggestion to replace

> <provider>org.hibernate.jpa.HibernatePersistenceProvider</provider>

in persistence.xml? Because I currently get that **org.hibernate.jpa.HibernatePersistenceProvider**  is not assignable to 'jakarta.persistence.spi.PersistenceProvider'

Best
A

**Comment by scottmarlow** [comment]
Hi @aligit 

The Hibernate ORM project does not yet produce builds with `jakarta.persistence.spi.PersistenceProvider`.  However, https://download.jboss.org/wildfly/22.0.0.Beta1/wildfly-preview-22.0.0.Beta1.zip `Jakarta EE 9` transformed Hibernate ORM jars.  

WildFly uses the https://github.com/wildfly-extras/batavia project to do the EE 9 transformation.  Batavia depends on https://github.com/eclipse/transformer along with EE 9 transformation rule property files that describe how to do the transformation.

Hope this helps.

Scott

**Pull request closed without merging**